examples/gno.land/p/moul/collection/collection.gno

```diff
@@ -0,0 +1,314 @@
+		}
+
+		// For non-unique indexes, we store the ID as the value
+		idx.tree.Set(key, idStr)
+	}
+
+	return uint64(id)
+}
+
+// Get retrieves an object by index and key, returns (object, id)
+func (c *Collection) Get(indexName, key string) (interface{}, uint64) {
+	idx, exists := c.indexes[indexName]
+	if !exists {
+		return nil, 0
+	}
+
+	if indexName == IDIndex {
+		obj, exists := idx.tree.Get(key)
+		if !exists {
+			return nil, 0
+		}
+		id, err := seqid.FromString(key)
+		if err != nil {
+			return nil, 0
+		}
+		return obj, uint64(id)
+	}
+
+	// For other indexes
+	if idx.options&CaseInsensitiveIndex != 0 {
+		key = strings.ToLower(key)
+	}
+
+	idStr, exists := idx.tree.Get(key)
+	if !exists {
+		return nil, 0
+	}
+
+	// Get the actual object from _id index
+	obj, exists := c.indexes[IDIndex].tree.Get(idStr.(string))
+	if !exists {
+		return nil, 0
+	}
+
+	id, err := seqid.FromString(idStr.(string))
+	if err != nil {
+		return nil, 0
+	}
+	return obj, uint64(id)
+}
+
+// GetIndex returns the underlying tree for an index
+func (c *Collection) GetIndex(name string) avl.ITree {
+	idx, exists := c.indexes[name]
+	if !exists {
+		return nil
+	}
+	return idx.tree
+}
+
+// Delete removes an object by its ID
+func (c *Collection) Delete(id uint64) {
+	idStr := seqid.ID(id).String()
+
+	// Get the object first to clean up other indexes
+	obj, exists := c.indexes[IDIndex].tree.Get(idStr)
+	if !exists {
+		return
+	}
+
+	// Remove from all indexes
+	for name, idx := range c.indexes {
+		if name == IDIndex {
+			idx.tree.Remove(idStr)
+			continue
+		}
+		key := idx.fn(obj)
+		idx.tree.Remove(key)
+	}
+}
+
+// Update updates an existing object and returns its ID (0 if not found)
+func (c *Collection) Update(id uint64, obj interface{}) uint64 {
+	if obj == nil {
+		return 0
+	}
+
+	idStr := seqid.ID(id).String()
+
+	// Check if object exists
+	oldObj, exists := c.indexes[IDIndex].tree.Get(idStr)
+	if !exists {
+		return 0
+	}
+
+	// Check uniqueness constraints
+	for name, idx := range c.indexes {
+		if name == IDIndex {
+			continue
+		}
+		if idx.options&UniqueIndex != 0 {
+			newKey, ok := safeGenerateKey(idx.fn, obj)
+			if !ok {
+				return 0
+			}
+			oldKey, ok := safeGenerateKey(idx.fn, oldObj)
+			if !ok {
+				return 0
+			}
+			// If the key changed and new key already exists
+			if newKey != oldKey {
+				if existing, _ := idx.tree.Get(newKey); existing != nil {
+					return 0 // Uniqueness constraint violated
+				}
+			}
+		}
+	}
+
+	// Remove old index entries
+	for name, idx := range c.indexes {
+		if name == IDIndex {
+			continue
+		}
+		oldKey, ok := safeGenerateKey(idx.fn, oldObj)
+		if !ok {
+			continue
+		}
+		idx.tree.Remove(oldKey)
+	}
+
+	// Add new index entries
+	for name, idx := range c.indexes {
+		if name == IDIndex {
+			idx.tree.Set(idStr, obj)
+			continue
+		}
+		newKey, ok := safeGenerateKey(idx.fn, obj)
+		if !ok {
+			// Rollback: restore old object
+			c.indexes[IDIndex].tree.Set(idStr, oldObj)
+			return 0
+		}
```

Comment:
When rollback occurs due to `safeGenerateKey` call failures or `UniqueIndex` conflicts, only the `_id` index of `oldObj` is currently being restored.

I didn't found a mechanism to restore previously removed entries in other indices. This can lead to inconsistency between the `_id` index and other indices, may potentially compromising data integrity.

Reply:
What do you think about fa46373e9?

Reply:
I like it 👍